File: javascript/src/problems301_400/decode-string.js
class Problem {
  decodeString(s) {
    throw new Error(`Not implemented (${s}).`);
  }
}

class Default extends Problem {
  decodeString(s) {
    const stack = [];
    for (const c of s) {
      if (c !== ']') {
        stack.push(c);
        continue;
      }
      let content = '';
      while (stack[stack.length - 1] !== '[') {
        content = stack.pop() + content;
      }
      stack.pop(); // remove '['
      let timesStr = '';
      while (stack.length > 0 && !isNaN(parseInt(stack[stack.length - 1]))) {
        timesStr = stack.pop() + timesStr;
      }
      const times = parseInt(timesStr);
      for (let i = 0; i < times; i++) {
        for (const contentC of content) {
          stack.push(contentC);
        }
      }
    }
    return stack.join('');
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems301_400/decode_string.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def decode_string(self, s: str) -> str:
        pass


class Default(Problem):
    @override
    def decode_string(self, s: str) -> str:
        stack: list[str] = []
        for char in s:
            if char != ']':
                stack.append(char)
                continue
            content_builder: str = ''
            time_builder: str = ''
            while stack[-1] != '[':
                content_builder += stack.pop()
            stack.pop()  # remove '['
            while stack and stack[-1].isdigit():
                time_builder += stack.pop()

            content = content_builder[::-1]
            times = int(time_builder[::-1])

            for _ in range(times):
                for content_char in content:
                    stack.append(content_char)

        return ''.join(stack)


SOLUTIONS: list[Problem] = [Default()]
